File: backend/routers/pdf.py
"""FastAPI routes for PDF operations."""

from fastapi import APIRouter, File, HTTPException, UploadFile
from fastapi.responses import FileResponse, Response
from pydantic import BaseModel

from services import pdf_ops

router = APIRouter(prefix="/api/pdf", tags=["pdf"])


# ── Upload ─────────────────────────────────────────────────────────────────────

@router.post("/upload")
async def upload_pdf(file: UploadFile = File(...)):
    if file.content_type != "application/pdf":
        raise HTTPException(400, "Only PDF files are accepted")
    data = await file.read()
    file_id = pdf_ops.save_upload(data)
    count = pdf_ops.page_count(file_id)
    return {"file_id": file_id, "page_count": count, "filename": file.filename}


# ── Preview ────────────────────────────────────────────────────────────────────

@router.get("/preview/{file_id}/{page}")
def get_page_preview(file_id: str, page: int):
    try:
        png = pdf_ops.render_page_thumbnail(file_id, page)
    except Exception as e:
        raise HTTPException(404, str(e))
    return Response(content=png, media_type="image/png")


# ── Merge ──────────────────────────────────────────────────────────────────────

class MergeRequest(BaseModel):
    file_ids: list[str]


@router.post("/merge")
def merge(req: MergeRequest):
    out_id = pdf_ops.merge_pdfs(req.file_ids)
    count = pdf_ops.page_count(out_id)
    return {"file_id": out_id, "page_count": count}


# ── Split ──────────────────────────────────────────────────────────────────────

class SplitRange(BaseModel):
    start: int
    end: int


class SplitRequest(BaseModel):
    file_id: str
    ranges: list[SplitRange]


@router.post("/split")
def split(req: SplitRequest):
    range_tuples = [(r.start, r.end) for r in req.ranges]
    out_ids = pdf_ops.split_pdf(req.file_id, range_tuples)
    return {"file_ids": out_ids}


# ── Rotate ─────────────────────────────────────────────────────────────────────

class RotateRequest(BaseModel):
    file_id: str
    rotations: dict[int, int]  # {page_index: degrees}


@router.post("/rotate")
def rotate(req: RotateRequest):
    out_id = pdf_ops.rotate_pages(req.file_id, req.rotations)
    count = pdf_ops.page_count(out_id)
    return {"file_id": out_id, "page_count": count}


# ── Reorder ────────────────────────────────────────────────────────────────────

class ReorderRequest(BaseModel):
    file_id: str
    order: list[int]


@router.post("/reorder")
def reorder(req: ReorderRequest):
    out_id = pdf_ops.reorder_pages(req.file_id, req.order)
    count = pdf_ops.page_count(out_id)
    return {"file_id": out_id, "page_count": count}


# ── Delete pages ───────────────────────────────────────────────────────────────

class DeleteRequest(BaseModel):
    file_id: str
    keep: list[int]


@router.post("/delete-pages")
def delete_pages(req: DeleteRequest):
    out_id = pdf_ops.delete_pages(req.file_id, req.keep)
    count = pdf_ops.page_count(out_id)
    return {"file_id": out_id, "page_count": count}


# ── Watermark ──────────────────────────────────────────────────────────────────

class WatermarkRequest(BaseModel):
    file_id: str
    text: str
    opacity: float = 0.3
    font_size: int = 48


@router.post("/watermark")
def watermark(req: WatermarkRequest):
    out_id = pdf_ops.add_watermark(req.file_id, req.text, req.opacity, req.font_size)
    count = pdf_ops.page_count(out_id)
    return {"file_id": out_id, "page_count": count}


# ── Download ───────────────────────────────────────────────────────────────────

@router.get("/download/{file_id}")
def download(file_id: str):
    path = pdf_ops.tmp_path(file_id)
    if not path.exists():
        raise HTTPException(404, "File not found")
    return FileResponse(str(path), media_type="application/pdf", filename="edited.pdf")


# ── Render without text (for editor background) ────────────────────────────────

@router.get("/render-notext/{file_id}/{page}")
async def render_no_text(file_id: str, page: int, scale: float = 1.8):
    try:
        data = pdf_ops.render_page_no_text(file_id, page, scale)
    except Exception as e:
        raise HTTPException(404, str(e))
    return Response(content=data, media_type="image/png")


# ── Page elements (for editor) ─────────────────────────────────────────────────

@router.get("/elements/{file_id}/{page}")
def get_elements(file_id: str, page: int):
    try:
        return pdf_ops.get_page_elements(file_id, page)
    except Exception as e:
        raise HTTPException(400, str(e))


# ── Apply in-page edits ────────────────────────────────────────────────────────

class ApplyEditsRequest(BaseModel):
    file_id: str
    page_index: int
    edits: list[dict]


@router.post("/apply-edits")
def apply_edits(req: ApplyEditsRequest):
    try:
        out_id, count = pdf_ops.apply_page_edits(req.file_id, req.page_index, req.edits)
        return {"file_id": out_id, "page_count": count}
    except Exception as e:
        raise HTTPException(400, str(e))

File: frontend/src/api/pdf.ts
const BASE = "http://localhost:8000/api/pdf";

export interface UploadResult {
  file_id: string;
  page_count: number;
  filename: string;
}

export async function uploadPDF(file: File): Promise<UploadResult> {
  const form = new FormData();
  form.append("file", file);
  const res = await fetch(`${BASE}/upload`, { method: "POST", body: form });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export function previewURL(file_id: string, page: number) {
  return `${BASE}/preview/${file_id}/${page}`;
}

export const noTextRenderURL = (fileId: string, page: number, scale = 1.8) =>
  `${BASE}/render-notext/${fileId}/${page}?scale=${scale}`;

export function downloadURL(file_id: string) {
  return `${BASE}/download/${file_id}`;
}

export async function mergePDFs(file_ids: string[]): Promise<{ file_id: string; page_count: number }> {
  const res = await fetch(`${BASE}/merge`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_ids }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export async function reorderPages(file_id: string, order: number[]): Promise<{ file_id: string; page_count: number }> {
  const res = await fetch(`${BASE}/reorder`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_id, order }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export async function deletePages(file_id: string, keep: number[]): Promise<{ file_id: string; page_count: number }> {
  const res = await fetch(`${BASE}/delete-pages`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_id, keep }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export async function rotatePages(
  file_id: string,
  rotations: Record<number, number>
): Promise<{ file_id: string; page_count: number }> {
  const res = await fetch(`${BASE}/rotate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_id, rotations }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export async function addWatermark(
  file_id: string,
  text: string,
  opacity = 0.3,
  font_size = 48
): Promise<{ file_id: string; page_count: number }> {
  const res = await fetch(`${BASE}/watermark`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_id, text, opacity, font_size }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export async function splitPDF(
  file_id: string,
  ranges: { start: number; end: number }[]
): Promise<{ file_ids: string[] }> {
  const res = await fetch(`${BASE}/split`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_id, ranges }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export interface PageElement {
  type: "text";
  x: number; y: number; w: number; h: number;
  text: string; fontSize: number; color: string;
  fromPdf: boolean;
}

export async function getPageElements(
  file_id: string,
  page: number
): Promise<{ elements: PageElement[]; pageWidth: number; pageHeight: number }> {
  const res = await fetch(`${BASE}/elements/${file_id}/${page}`);
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

export async function applyEdits(
  file_id: string,
  page_index: number,
  edits: object[]
): Promise<{ file_id: string; page_count: number }> {
  const res = await fetch(`${BASE}/apply-edits`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_id, page_index, edits }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}
